core/model/impl/src/main/kotlin/parsing/BooleanExpressionParser.kt

```diff
@@ -0,0 +1,302 @@
+
+        return inversePostfix
+    }
+
+    private fun buildExpressionTree(inversePostfix: List<Token>): E {
+        val expressionStack = arrayListOf<E>()
+        for (token in inversePostfix) {
+            when (token.kind) {
+                UnaryNot -> expressionStack.add(factory.createNot(expressionStack.removeLast()))
+                BinaryAnd -> expressionStack.add(factory.createAnd(
+                    rhs = expressionStack.removeLast(),
+                    lhs = expressionStack.removeLast(),
+                ))
+
+                BinaryOr -> expressionStack.add(factory.createOr(
+                    rhs = expressionStack.removeLast(),
+                    lhs = expressionStack.removeLast(),
+                ))
+
+                Variable -> {
+                    when (val result = factory.parseVariable(token.text)) {
+                        is Factory.ParseResult.Error -> throw ParseException(
+                            source = expressionSource,
+                            where = token.sourceLocation..<token.sourceLocation + token.text.length - 1,
+                            what = "Variable: ${result.message}"
+                        )
+
+                        is Factory.ParseResult.Ok ->
+                            expressionStack.add(result.variable)
+                    }
+                }
+
+                else -> throw AssertionError("Unexpected token in postfix notation - $token")
+            }
+        }
+        return expressionStack.single()
+    }
+
+    private data class Token(
+        val kind: Kind,
+        val sourceLocation: Int,
+        val text: String,
+    ) {
+        enum class Kind(val description: String) {
+            BeginOfExpression("<begin-of-expression>"),
+            LeftParenthesis("'('"),
+            RightParenthesis("')'"),
+            UnaryNot("'!'"),
+            BinaryAnd("'&'"),
+            BinaryOr("'|'"),
+            Variable("<variable>"),
+            EndOfExpression("<end-of-expression>"),
+        }
+
+        override fun toString(): String {
+            return when (kind) {
+                Variable -> "<variable '$text'>"
+                else -> kind.description
+            }
+        }
+    }
+
+    companion object {
+        private val nextToken: Map<Token.Kind, Set<Token.Kind>> = enumMapOf(
+            BeginOfExpression to enumSetOf(
+                Variable, UnaryNot, LeftParenthesis,
+            ),
+            LeftParenthesis to enumSetOf(
+                Variable, UnaryNot, LeftParenthesis,
+            ),
+            RightParenthesis to enumSetOf(
+                BinaryOr, BinaryAnd, RightParenthesis, EndOfExpression,
+            ),
+            BinaryAnd to enumSetOf(
+                Variable, UnaryNot, LeftParenthesis,
+            ),
+            BinaryOr to enumSetOf(
+                Variable, UnaryNot, LeftParenthesis,
+            ),
+            UnaryNot to enumSetOf(
+                Variable, UnaryNot, LeftParenthesis,
```

Comment:
Should multiple negations be allowed? Seems that increases mistake probability...

Reply:
That's an interesting question. Strictly speaking, this `next(token)` function is not something arbitrary, but is actually built from the expression grammar. And in grammar, we have rules like `<expression> ::= '!' <expression> | ...`. So there's no way to forbid that inner expression from being a negation itself on the grammar level in context-free fashion. However, what you are suggesting to do with the `next()` function should  actually "just work", but I don't really want to do it on the syntax level - it'll be a bit unnatural and confusing.

Consider also things like `"!(!a)"` which is still a double negation, but could not be possibly forbidden on the grammar level. One could argue, that if user writes something like that, then they explicitly want to have a double-negation, though still there's no reason to have that.

So it could be a warning on the semantic level (when expression tree is built). Maybe I'll add it later if this new API proves successful.

Thanks for the suggestion anyway!